### PR TITLE
Implement user provided aggregation specs

### LINKED ISSUE #360: Allow for user-provided aggregation
### Current and desired situation

Aggregation on group levels is specified in dictionaries within the python files of the code base. E.g. 

```python
aggregation_kindergeld = {
    "kumulativer_kindergeld_anspruch_tu": {
        "source_col": "kindergeld_anspruch",
        "aggr": "cumsum",
    },
    "anz_kinder_mit_kindergeld_tu": {
        "source_col": "kindergeld_anspruch",
        "aggr": "sum",
    },
}
```

- For user-written functions, it is currently impossible to specify aggregation variables in a similar way.

### Proposed implementation

- Add a new argument `aggregation_specs` to `compute_taxes_and_transfers`. 
- Elements of this dictionary are added to the aggregation specifications specified in the code base and override those if a key is used twice.

### Considered alternatives

- Add it to the already existing argument `functions`. We then would need to differentiate between a dictionary of functions and a dictionary of aggregation specifications.

### PR DESCRIPTION
**Pull request by ChristianZimpelmann**
### What problem do you want to solve?

- For user-written functions, it is currently impossible to specify aggregation variables in a similar way.

### Proposed implementation

- Add a new argument `aggregation_specs` to `compute_taxes_and_transfers`. 
- Elements of this dictionary are added to the aggregation specifications specified in the code base and override those if a key is used twice.

Closes #360 

### Todo

- [x] Implement new argument for `compute_taxes_and_transfers` as described in #360.
- [x] Write tests
- [x] Mention in tutorials

### CODE REVIEW

**Comment by hmgaudecker** [comment]
Running advanced usage tutorial fails:

```pytb
[File /mnt/econ/gettsim/gettsim/gettsim/visualization.py:133, in plot_dag(functions, targets, columns_overriding_functions, check_minimal_specification, selectors, labels, tooltips, plot_kwargs, arrow_kwargs, edge_kwargs, label_kwargs, node_kwargs)
    ]()[131](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=130)[ # Create and add aggregation functions
    ]()[132](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=131)[ typical_data_cols = list(TYPES_INPUT_VARIABLES)
--> ]()[133](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=132)[ aggregation_funcs = _create_aggregation_functions(
    ]()[134](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=133)[     user_and_internal_functions, targets, typical_data_cols
    ]()[135](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=134)[ )
    ]()[136](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=135)[ all_functions = {**user_and_internal_functions, **aggregation_funcs}
    ]()[138](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=137)[ all_functions = {
    ]()[139](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=138)[     k: v for k, v in all_functions.items() if k not in columns_overriding_functions
    ]()[140](file:///mnt/econ/gettsim/gettsim/gettsim/visualization.py?line=139)[ }

TypeError: _create_aggregation_functions() missing 1 required positional argument: 'user_provided_aggregation_specs']()
```

Aside: That was a misunderstanding/incomplete discussion yesterday re the basic usage. I'd prefer two households, maybe one single andone couple? The data should have the same form as in an actual application (multiple households).

File: gettsim/interface.py [hunk 1]
```diff
@@ -54,13 +55,16 @@ def compute_taxes_and_transfers(
...
-        Function from the policy environment. Functions can be anything of the specified
-        types and a list of the same objects. If the object is a dictionary, the keys of
-        the dictionary are used as a name instead of the function name. For all other
-        objects, the name is inferred from the function name.
+        Function from the policy environment. Functions can be anything of the
+        specified types and a list of the same objects. If the object is a dictionary,
+        the keys of the dictionary are used as a name instead of the function name. For
+        all other objects, the name is inferred from the function name.
+    aggregation_specs : dict, default None
+        A dictionary which contains specs for functions which aggregate variables on
```
Comment (hmgaudecker) [inline gettsim/interface.py:66]: Refer to docs?

File: gettsim/interface.py [hunk 2]
```diff
@@ -152,6 +168,9 @@ def check_data_check_functions_and_merge_functions(
         needed by the user.
     data : dict of pandas.Series
         Data provided by the user.
+    aggregation_specs : dict, default None
+        A dictionary which contains specs for functions which aggregate variables on
```
Comment (hmgaudecker) [inline gettsim/interface.py:172]: Again, refer to some place where I find useful information how this should look like. If I am lost before, I am also lost after reading this.

File: gettsim/interface.py [hunk 3]
```diff
@@ -815,7 +836,14 @@ def _create_aggregation_functions(user_and_internal_functions, targets, data_col
         agg_col: {"aggr": "sum", "source_col": rchop(rchop(agg_col, "_tu"), "_hh")}
         for agg_col in automated_sum_aggregation_cols
     }
-    aggregation_dict = {**aggregation_dict, **automated_sum_aggregation_specs}
+
+    # Add automated aggregation specs.
+    # Note: For duplicate keys, automated specs are treated with lower priority.
```
Comment (hmgaudecker) [inline gettsim/interface.py:841]:
```suggestion
    # Note: For duplicate keys, explicitly set specs are treated with higher priority than automated specs.
```
Better the positive version. More importantly, keep it the same way as next docstring.

File: gettsim/parameters/soz_vers_beitr.yaml [hunk 4]
```diff
@@ -783,10 +783,10 @@ ag_abgaben_geringf:
...
 
-ges_pflegev_zusatz_kinderlos_altersgrenze:
+ges_pflegev_zusatz_kinderlos_mindestalter:
   name:
-    de: Altersgrenze Beitragszuschlag für Kinderlose
-    en: Age limit additional contribution for childless
+    de: Mindestalter Beitragszuschlag für Kinderlose
+    en: Minimum age for additional contribution for childless
   description:
     de: Der Beitragssatz für Kinderlose steigt ab dieser Altersgrenze an.
```
Comment (hmgaudecker) [inline gettsim/parameters/soz_vers_beitr.yaml:791]:
```suggestion
    de: Der Beitragssatz für Kinderlose ist ab diesem Alter höher.
```

File: gettsim/parameters/soz_vers_beitr.yaml [hunk 5]
```diff
@@ -783,10 +783,10 @@ ag_abgaben_geringf:
...
     en: The contribution rate for childless people increases from this age limit.
```
Comment (hmgaudecker) [inline gettsim/parameters/soz_vers_beitr.yaml:792]:
```suggestion
    en: Childless individuals face a higher contribution rate starting at this age.
```

File: gettsim/taxes/zu_verst_eink/vorsorgeaufw.py [hunk 6]
```diff
@@ -254,7 +254,7 @@ def _vorsorgeaufw_vom_lohn_bis_2019_single(
     bruttolohn_m: float,
     eink_st_abzüge_params: dict,
 ) -> float:
-    """Calcaulate vorsoge expenditures until 2019 for singles.
+    """Calculate vorsoge expenditures until 2019 for singles.
```
Comment (hmgaudecker) [inline gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:257]:
```suggestion
    """Calculate precautionary expenditures until 2019 for singles.
```
or "Vorsorge", but capitalised and without typo :smile:, same below.

File: gettsim/taxes/zu_verst_eink/vorsorgeaufw.py [hunk 7]
```diff
@@ -280,7 +280,7 @@ def _vorsorgeaufw_vom_lohn_bis_2019_tu(
     bruttolohn_m_tu: float,
     eink_st_abzüge_params: dict,
 ) -> float:
-    """Calcaulate vorsoge expenditures until 2019 per tax unit.
+    """Calculate vorsoge expenditures until 2019 per tax unit.
```
Comment (hmgaudecker) [inline gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:283]:
```suggestion
    """Calculate precautionary expenditures until 2019 per tax unit.
```

File: gettsim/taxes/zu_verst_eink/vorsorgeaufw.py [hunk 8]
```diff
@@ -310,7 +310,7 @@ def _berechne_vorsorgeaufw_bis_2004(
     anzahl_erwachsene: int,
     eink_st_abzüge_params: dict,
 ) -> float:
-    """Calcaulate vorsoge expenditures until 2004.
+    """Calculate vorsoge expenditures until 2004.
```
Comment (hmgaudecker) [inline gettsim/taxes/zu_verst_eink/vorsorgeaufw.py:313]:
```suggestion
    """Calculate precautionary expenditures until 2004.
```

**Review by hmgaudecker — COMMENTED** [review]
Excellent, thanks!!! 

Very minor stuff, except for the missing argument noted elsewhere.

**Comment by ChristianZimpelmann** [comment]
Many thanks for the suggestions!

- I fixed the error in visualizations (and added a test for it). Advanced usage is running through now.
- I added a second household to basic usage.

**Comment by hmgaudecker** [comment]
Somewhere the behaviour got lost that we show all missing columns, can we restore that?

image

(from advanced tutorial, else would need to adjust that)

**Comment by hmgaudecker** [comment]
I fixed the issues in #365 in the previous commit. The notebooks `policy_functions` and `parameters` currently fail. I have no idea what somebody tried to do there, should never have been publicly visible. Please fix that, I won't manage today (maybe tomorrow).

Re `synthetic.py`: This is a good example of code that would work well to cook up a one-time project. It does not live up to standards for a long-term collaborative project (clear API, declarative instead of imperative code, ...). For the future: Please ensure quality standards before merging. Better to have a missing feature than an enduring nightmare. Sorry for not looking into it back then. Thanks.

**Comment by ChristianZimpelmann** [comment]
- I adjusted the order of input data checks
- uploading code coverage reports is back in (got lost when commenting out jax tests)
- I fixed the problem in `synthetic.py` that made the tutorials fail (thx for fixing the issue)

btw.: Is there a better way to check the tutorials other than selecting each one and clicking "Run all Cells"?

**Comment by hmgaudecker** [comment]
Thanks! 

> btw.: Is there a better way to check the tutorials other than selecting each one and clicking "Run all Cells"?

Should be possible to let them run as part of the CI or so, not sure. Only problem would be with the advanced_usage where we show errors, but maybe one could continue with execution despite them.

**Review by hmgaudecker — APPROVED** [review]
Wonderful, thanks!

**Pull request merged**